azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/sync_clients.py

```diff
@@ -1,24 +1,22 @@
-# --------------------------------------------------------------------------------------------
-# Copyright (c) Microsoft Corporation. All rights reserved.
-# Licensed under the MIT License. See License.txt in the project root for license information.
-# --------------------------------------------------------------------------------------------
-
 import logging
+import abc
+import six
 from threading import Event
-from .transport.mqtt.mqtt_transport import MQTTTransport
+from .transport.mqtt import MQTTTransport
 
 logger = logging.getLogger(__name__)
 
 
-class InternalClient(object):
+@six.add_metaclass(abc.ABCMeta)
```

Comment:
I defined this as an abstract class, which I think might kind of be overkill. My reasoning was, GenericClient shouldn't ever be instantiated, and by defining abstract methods (connect, etc.) we could always guarantee that there was parity between Sync and Async versions of a given method - but only if an abstract method was defined. Not really sure how common it would be for someone to remember to define the abstract method, but then forget to do one of the sync/async ones.

Additionally, the subclasses, GenericSyncClient and GenericAsyncClient are also not supposed to be instantiated, but there's no way to stop them from being, given that there are Device and Module specific methods (i.e. c2d) on the horizon.

I still went ahead and used the abstract base class, but depending on your thoughts, I may remove it - I don't want to add something just for the sake of it, especially when it A) nets us very little, and B) still will require us to fall back on convention. Perhaps convention alone would be sufficient.

Reply:
Not overkill.  I think that forgetting to implement either sync or async functions could be super common.  Any time a developer is required to do the same (or mostly same) thing in more than one place, someone is going to mess it up unless you enforce it.  It's like having a required parameter but not enforcing it.

---
In reply to: [246596935](https://github.com/Azure/azure-iot-sdk-python-preview/pull/39#discussion_r246596935) [](ancestors = 246596935)

Reply:
However, if we opt for async callbacks, then we might not need generic clients because sharing callback code seems to be most of the justification for having these classes.

---
In reply to: [246948932](https://github.com/Azure/azure-iot-sdk-python-preview/pull/39#discussion_r246948932) [](ancestors = 246948932,246596935)

Reply:
Yup. I didn't want to touch the callbacks yet since they deal with the transport, and the implications there get a bit more nebulous to me. We would still need the GenericSync / GenericAsync clients however for shared Device/Module functionality

---
In reply to: [246953315](https://github.com/Azure/azure-iot-sdk-python-preview/pull/39#discussion_r246953315) [](ancestors = 246953315,246948932,246596935)